sdks/java/core/src/main/java/org/apache/beam/sdk/runners/TransformHierarchy.java

```diff
@@ -506,30 +506,55 @@ void replaceOutputs(Map<PValue, ReplacementOutput> originalToReplacement) {
      * <p>Provides an ordered visit of the input values, the primitive transform (or child nodes for
      * composite transforms), then the output values.
      */
-    private void visit(PipelineVisitor visitor, Set<PValue> visitedValues) {
+    private void visit(
+        PipelineVisitor visitor,
+        Set<PValue> visitedValues,
+        Set<Node> visitedNodes,
+        Set<Node> passedComposites) {
+      if (getEnclosingNode() != null && !visitedNodes.contains(getEnclosingNode())) {
+        getEnclosingNode().visit(visitor, visitedValues, visitedNodes, passedComposites);
```

Comment:
Assuming our other invariants hold, we'll walk up the stack of nodes, and then immediately return from all of them. We could short-circuit this check in the case we have already been visited, but doing so requires duplicating rather than moving the code (as we may/will visit this node as part of the process of visiting its parent)